image-builder: create directory

supakeen opened the pull request:
Create the output directory; this is a no-op if it already exists but fixes #94 if it doesn't. The code is similar to (but not identical to) the checks done in `cmd/image-builder/manifest.go` for the `withSBOM` parts.

We might (or I) might want to consolidate those two codepaths either here in a followup.

mvo5 commented on cmd/image-builder/build.go:
```diff
@@ -26,9 +26,21 @@ func buildImage(res *imagefilter.Result, osbuildManifest []byte, opts *buildOpti
+
 	if opts.WriteManifest {
-		p := filepath.Join(opts.OutputDir, fmt.Sprintf("%s.osbuild-manifest.json", outputNameFor(res)))
-		if err := os.WriteFile(p, osbuildManifest, 0644); err != nil {
```
Thanks for looking into this and the patch!

Would this:
```go
diff --git a/cmd/image-builder/build.go b/cmd/image-builder/build.go
index 26b6151..b00c9c6 100644
--- a/cmd/image-builder/build.go
+++ b/cmd/image-builder/build.go
@@ -28,6 +28,9 @@ func buildImage(res *imagefilter.Result, osbuildManifest []byte, opts *buildOpti
        }
        if opts.WriteManifest {
                p := filepath.Join(opts.OutputDir, fmt.Sprintf("%s.osbuild-manifest.json", outputNameFor(res)))
+               if err := os.MkdirAll(filepath.Dir(p), 0755); err != nil {
+                       return err
+               }
                if err := os.WriteFile(p, osbuildManifest, 0644); err != nil {
                        return err
                }
```
also fix the issue? Having a small test (or modifiying one of the existing ones) to trigger this issue would be great.

mvo5 replied:
Thanks, I have no strong opinion, the two are slightly different but maybe:
```go
diff --git a/cmd/image-builder/manifest.go b/cmd/image-builder/manifest.go
index 7a4ab0c..ac60674 100644
--- a/cmd/image-builder/manifest.go
+++ b/cmd/image-builder/manifest.go
@@ -24,6 +24,9 @@ type manifestOptions struct {
 
 func sbomWriter(outputDir, filename string, content io.Reader) error {
        p := filepath.Join(outputDir, filename)
+       if err := os.MkdirAll(filepath.Dir(p), 0755); err != nil {
+               return err
+       }
        f, err := os.Create(p)
        if err != nil {
                return err
@@ -50,9 +53,6 @@ func generateManifest(dataDir string, img *imagefilter.Result, output io.Writer,
                if outputDir == "" {
                        outputDir = outputNameFor(img)
                }
-               if err := os.MkdirAll(outputDir, 0755); err != nil {
-                       return err
-               }
                manifestGenOpts.SBOMWriter = func(filename string, content io.Reader) error {
                        return sbomWriter(outputDir, filename, content)
                }

```
makes sense, would make the sbomWriter easier to reuse.

supakeen commented:
@mvo5 I've addressed the testcase, I moved some commits around so the testcase is first (so it fails *before* the rest of the commits are applied). The testcase is in the integration tests.

Aside from that I've evened out the approaches taken in both `sbomWriter` and `buildImage`.

mvo5 commented on test/test_container.py:
```diff
@@ -52,3 +52,23 @@ def test_container_manifest_generates_sbom(tmp_path, build_container):
+        "--privileged",
+        "-v", f"{output_dir}:/output",
+        build_container,
+        "build",
```
Ideally we would use the "build_fake_container" here that mocks osbuild (c.f. https://github.com/osbuild/image-builder-cli/pull/21/commits/13279a2562eb3b4b3ea0932c4ca0c322f3baa12a) as for this particular test we don't really need the full build. But its fine, we can tweak it after the other one is merged (or if you prefer you could cherry pick that commit and replace "build_container" with "build_fake_container" in the test and it should run a lot quicker :)

supakeen replied:
I think we do; because if we don't then the actual error (that was there) isn't triggered? Perhaps if we mock the filesystem as well; but that'd be a followup.

mvo5 reviewed (approved):
Thank you, this looks very nice now and extra thanks for the test. One tiny suggestion inline but that can be a followup
